### ISSUE TITLE
Wildcard retrieval of files

### ISSUE THREAD

**Issue by nicholas-miles** [awslabs/aws-data-wrangler#322]
I am attempting to access data in our s3 datalake. Since production systems are writing the data I want along with other files that I don't want, simply using a prefix is insufficient to get the data I need.

Dask allows this sort of wildcarding, for example 
```
import dask.dataframe as dd
dd.read_parquet(f's3://{bucket}/prod-system-*/*/parquet/*.parquet')
```

Using awswrangler for the above task isn't viable. I know that boto3 doesn't allow for wildcard filtering, but surely it must be doable if dask is able to implement that functionality?

**Comment by nicholas-miles** [comment]
It appears that dask uses [`s3fs`](https://s3fs.readthedocs.io/en/latest/api.html) to do the wildcard matching. Happy to build it myself for my purposes, but perhaps that would be a good inclusion in this package?

**Comment by nicholas-miles** [comment]
This works!

```
datalake = 'datalake'
b3 = boto3.Session(profile_name=datalake)
fs = s3fs.S3FileSystem(anon=False, profile=datalake)
    
def retrieve(self, url, format='csv') -> pd.DataFrame:
        if isinstance(url, list):
            files = url
        else:
            files = fs.glob(url)

        if format == 'csv':
            data = wr.s3.read_csv(files, boto3_session=b3)
        elif format == 'parquet':
            data = wr.s3.read_parquet(files, boto3_session=b3)

        return data
```

**Comment by igorborgest** [comment]
Hi @nicholas-miles,

For sure is a nice feature!

Unfortunately it will not be so straightforward to include cause now we don't use `s3fs` to list files on s3. We are only using it during reads/writes. But I think we can adapt our current code to support it even w/o `s3fs`.

Please feel free to send a PR if you want, if not I will try to address that myself further.

Thanks!

**Comment by alan-aipe** [comment]
Hi @igorborgest ,
May I take this task up if no one else is already working on it? I would like to start contributing to repo and this seems like a good first issue

**Comment by igorborgest** [comment]
Hi @alan-aipe!

Yeah, good first issue indeed. I will assign you on it! 🚀

**Comment by alan-aipe** [comment]
Sorry @igorborgest . I am unassigning myself from this because I got a heavy task today from my team. Therefore, I wont be able to contribute for next 3-4 weeks. Please re-assign. I will pick up a new task once I am free. I apologize again for inconvenience caused.

**Comment by nicholas-miles** [comment]
@igorborgest I'm happy to take it on! I'd love any guidance you can give on how to do this though. I'm assuming we want to avoid additional imports/dependencies

**Comment by igorborgest** [comment]
@nicholas-miles yes, we don't want new dependencies, and I don't think it is necessary here. Basically we just need to adapt the internal [`_path2list()`](https://github.com/awslabs/aws-data-wrangler/blob/fb1615c20288be1ab1e9d1493306d4957e4c8446/awswrangler/s3/_list.py#L59) that converted all `path` arguments spread around the code base to a list of file paths.

I think we should start simple and create a function that will filter a list of paths based on wildcards. Then would be only a matter to split the received path just before the first * occurrence, let this stripped prefix follows the regular function flow and in the end insert the filter function created just before the return point.

Does it make sense?

**Comment by igorborgest** [comment]
The filter function itself could be done using this [fnmatch approach](https://stackoverflow.com/a/11427183/4162047).

[fnmatch](https://docs.python.org/3/library/fnmatch.html) would be great because it will allow us to use others cool unix patterns like: `?`, `[seq]`, `[seq]`.

**Comment by igorborgest** [comment]
So, what do you think @nicholas-miles? Can I assign you on it?

**Comment by nicholas-miles** [comment]
Sure! I'll have a look this weekend.

**Comment by igorborgest** [comment]
@nicholas-miles thanks for your contribution.

We already have this feature done to release. Could you test it before the official release?
The ideia is publish the feature in the version `1.8.0` next weekend.

```sh
pip install git+https://github.com/awslabs/aws-data-wrangler.git@dev
```

Basically all function in the library that receives an S3 prefix as argument now supports Unix style wildcards.

**Comment by igorborgest** [comment]
Hi @nicholas-miles !

It is totally off-topic, but we are stating a [_"Who uses AWS Data Wrangler?"_](https://github.com/awslabs/aws-data-wrangler/tree/dev#who-uses-aws-data-wrangler) section. So feel free to add yourself if you want 😄 .

**Comment by igorborgest** [comment]
Released on version [1.8.0](https://github.com/awslabs/aws-data-wrangler/releases/tag/1.8.0).

**Comment by tochandrashekhar** [comment]
I am still not able to use wildcard for wr.s3.list_objects. Can you please provide some example?
